src/platform/backends/qemu/qemu_virtual_machine.cpp

```diff
@@ -343,31 +344,57 @@ void mp::QemuVirtualMachine::start()
     vm_process->write(qmp_execute_json("qmp_capabilities"));
 }
 
-void mp::QemuVirtualMachine::shutdown()
+void mp::QemuVirtualMachine::shutdown(const bool force)
 {
-    if (state == State::suspended)
+    std::unique_lock<std::mutex> lock{state_mutex};
+
+    try
     {
-        mpl::log(mpl::Level::info, vm_name, fmt::format("Ignoring shutdown issued while suspended"));
+        check_state_for_shutdown(force);
     }
-    else
+    catch (const VMStateIdempotentException& e)
     {
-        drop_ssh_session();
+        mpl::log(mpl::Level::info, vm_name, e.what());
+        return;
+    }
+
+    if (force)
+    {
+        mpl::log(mpl::Level::info, vm_name, "Forcing shutdown");
 
-        if ((state == State::running || state == State::delayed_shutdown || state == State::unknown) && vm_process &&
-            vm_process->running())
+        if (vm_process)
         {
-            vm_process->write(qmp_execute_json("system_powerdown"));
-            vm_process->wait_for_finished(timeout);
+            mpl::log(mpl::Level::info, vm_name, "Killing process");
+            lock.unlock();
+            vm_process->kill();
+            if (vm_process != nullptr && !vm_process->wait_for_finished(timeout))
+            {
+                mpl::log(mpl::Level::error, vm_name, "Killing qemu process could not finish.");
```

Comment:
Hmm, why a log? This is an error and should I think it should fail the request, so I would expect this to throw an exception, as usual.

Also, could the message say instead "The QEMU process did not finish within x seconds"?

Reply:
Indeed, an exception is more appropriate here. About the text message, what about "The QEMU process **kill** did not finish within x seconds"? Which is more informative.

Reply:
Hmm, sending the "kill" signal itself finishes, this would be the process (or OS) not reacting on time. How about "The QEMU process did not finish within x seconds after being killed"?

Reply:
This text "The QEMU process did not finish within x seconds after being killed" is fine by me as well.